pkg/theiactl/portforwarder/portforwarder.go

```diff
@@ -0,0 +1,189 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package portforwarder
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+	"k8s.io/klog/v2"
+)
+
+type PortForwarder struct {
+	config        *rest.Config
+	clientset     kubernetes.Interface
+	namespace     string
+	name          string
+	targetPort    int
+	listenAddress string
+	listenPort    int
+	stopCh        chan struct{}
+}
+
+// This function creates Port Forwarder for a Pod
+// After creating Port Forwarder object, call Start() on it to start forwarding
+// channel and Stop() to terminate it
+func NewPortForwarder(config *rest.Config, namespace string, pod string, targetPort int, listenAddress string, listenPort int) (*PortForwarder, error) {
+	klog.Infof("Port forwarder requested for pod %s/%s: %s:%d -> %d", namespace, pod, listenAddress, listenPort, targetPort)
+
+	pf := &PortForwarder{
+		config:        config,
+		namespace:     namespace,
+		name:          pod,
+		targetPort:    targetPort,
+		listenAddress: listenAddress,
+		listenPort:    listenPort,
+	}
+
+	var err error
+	pf.clientset, err = kubernetes.NewForConfig(pf.config)
+	if err != nil {
+		return pf, fmt.Errorf("could not create kubernetes client: %v", err)
+	}
+
+	return pf, nil
+}
+
+// This function creates Port Forwarder for a Service by finding first Pod
+// that belongs to the Service, and target Port for requested Service Port.
+// This code is based upon kubectl port-forward implementation
+// After creating Port Forwarder object, call Start() on it to start forwarding
+// channel and Stop() to terminate it
+func NewServicePortForwarder(config *rest.Config, namespace string, service string, servicePort int, listenAddress string, listenPort int) (*PortForwarder, error) {
+	pf := &PortForwarder{
+		config:        config,
+		namespace:     namespace,
+		listenAddress: listenAddress,
+		listenPort:    listenPort,
+	}
+
+	var err error
+	pf.clientset, err = kubernetes.NewForConfig(pf.config)
+	if err != nil {
+		return pf, fmt.Errorf("could not create kubernetes client: %v", err)
+	}
+
+	serviceObj, err := pf.clientset.CoreV1().Services(pf.namespace).Get(context.TODO(), service, metav1.GetOptions{})
+	if err != nil {
+		return pf, fmt.Errorf("failed to read Service %s: %v", service, err)
+	}
+
+	// find container port that corresponds to requested service port
+	pf.targetPort, err = getContainerPortByServicePort(serviceObj, servicePort)
+	if err != nil {
+		return pf, err
+	}
+
+	klog.Infof("Port forwarder requested for service %s/%s: %s:%d -> %d", namespace, service, listenAddress, servicePort, pf.targetPort)
```

Comment:
I understand the advantage of logs. But as theiactl uses this package to do port forwarding and displays status of spark job or yaml format results. It seems a little bit strange to have the log before the results, and this exposes unnecessary details for CTL users. Would it be better to only have this message when there is an error, or in a debug mode?

Reply:
Agreed, I would only expose this in verbose mode. I don't think initial theiactl PR includes verbose mode, but once its added we can involve it here.